ingot/src/icmp.rs

```diff
@@ -3,22 +3,195 @@
 // file, You can obtain one at https://mozilla.org/MPL/2.0/.
 
 use ingot_macros::Ingot;
-use ingot_types::primitives::u16be;
+use ingot_types::{primitives::*, Vec};
+
+ingot_types::zerocopy_type!(
+    #[derive(Default)]
+    pub struct IcmpV4Type(pub u8)
+);
+
+impl IcmpV4Type {
+    pub const ECHO_REPLY: Self = Self(0);
+    pub const DESTINATION_UNREACHABLE: Self = Self(3);
+    pub const SOURCE_QUENCH: Self = Self(4);
+    pub const REDIRECT: Self = Self(5);
+    pub const ECHO_REQUEST: Self = Self(8);
+    pub const ROUTER_ADVERTISEMENT: Self = Self(9);
+    pub const ROUTER_SOLICITATION: Self = Self(10);
+    pub const TIME_EXCEEDED: Self = Self(11);
+    pub const PARAMETER_PROBLEM: Self = Self(12);
+    pub const TIMESTAMP: Self = Self(13);
+    pub const TIMESTAMP_REPLY: Self = Self(14);
+
+    /// This ICMP packet's payload contains a portion of the packet it was
+    /// generated in response to.
+    ///
+    /// This consists of the IPv4 header and 8 bytes of the upper protocol.
+    pub const fn payload_is_packet(self) -> bool {
+        matches!(
+            self,
+            Self::DESTINATION_UNREACHABLE
```

Comment:
Thanks for covering the additional cases beyond DU here. Time exceeded will be important for traceroute to work properly.